meteor/sofie-server/data-access/repositories/cache/cached-rundown-repository.ts

```diff
@@ -0,0 +1,48 @@
+import { RundownRepository } from '../interfaces/rundown-repository'
+import { Rundown } from '../../../model/entities/rundown'
+import { MisconfigurationException } from '../../../model/exceptions/misconfiguration-exception'
+import { Identifier } from '../../../model/interfaces/identifier'
+
+export class CachedRundownRepository implements RundownRepository {
+	private static instance: RundownRepository
+
+	public static getInstance(rundownRepository?: RundownRepository): RundownRepository {
+		if (!this.instance) {
+			if (!rundownRepository) {
+				throw new MisconfigurationException(
+					`No RundownRepository provided. Unable to create instance of ${CachedRundownRepository.name}`
+				)
+			}
+			this.instance = new CachedRundownRepository(rundownRepository)
+		}
+		return this.instance
+	}
+
+	private rundownRepository: RundownRepository
+
+	private cachedRundowns: Map<string, Rundown> = new Map()
+
+	constructor(rundownRepository: RundownRepository) {
+		this.rundownRepository = rundownRepository
+	}
+
+	public async getRundown(rundownId: string): Promise<Rundown> {
+		if (!this.cachedRundowns.has(rundownId)) {
+			console.log(`### Rundown with id: "${rundownId}" not found in cache. Loading rundown from database...`)
+			const rundown: Rundown = await this.rundownRepository.getRundown(rundownId)
+			this.cachedRundowns.set(rundownId, rundown)
+		}
+		return this.cachedRundowns.get(rundownId) as Rundown
```

Comment:
The problem is that `Map.get()` returns `V | undefined`. 
We know that the result is not undefined since we already checked for that earlier in the method.

Changing the return type of the method to `Rundown | undefined` is not an option since returning undefined or null is a bad practice.

Alternatively we could write something like:
```
public async getRundown(rundownId: string): Promise<Rundown> {
  if (!this.cachedRundowns.has(rundownId)) {
    // fetch rundown and put it in cache
  }
  const rundown: Rundown | undefined =  this.cachedRundowns.get(rundownId)
  if (!rundown)  {
    // throw Error
  }
  return rundown
}
```
And in that case I am more in favour of `as Rundown`

Reply:
I would write it as:
```
public async getRundown(rundownId: string): Promise<Rundown> {
  let rundown = this.cachedRundowns.get(rundownId)
  if (!rundown) {
    // fetch rundown and put it in cache
    rundown = .....
  }
  return rundown
}
```
or
```
public async getRundown(rundownId: string): Promise<Rundown> {
  let rundown = this.cachedRundowns.get(rundownId)
  if (rundown) return rundown

  // fetch rundown and put it in cache
}
```

Trust typescript to handle the `undefined` instead of fighting it. And this way it wont need to do two lookups in the Map when fetching a value.